src/aspire/estimation/covar2d.py

```diff
@@ -274,28 +285,30 @@ def get_cwf_coeffs(self, coeffs, ctf_fb=None, ctf_idx=None, mean_coeff=None, cov
         if covar_coeff is None:
             covar_coeff = self.get_covar(coeffs, ctf_fb, ctf_idx, mean_coeff, noise_var=noise_var)
 
+        # should be none or both
         if (ctf_fb is None) or (ctf_idx is None):
-            ctf_idx = np.zeros(coeffs.shape[1], dtype=int)
+            ctf_idx = np.zeros(coeffs.shape[0], dtype=int)
             ctf_fb = [BlkDiagMatrix.eye_like(covar_coeff)]
 
         noise_covar_coeff = noise_var * BlkDiagMatrix.eye_like(covar_coeff)
 
         coeffs_est = np.zeros_like(coeffs)
 
         for k in np.unique(ctf_idx[:]):
-            coeff_k = coeffs[:, ctf_idx == k]
+            coeff_k = coeffs[ctf_idx == k]
             ctf_fb_k = ctf_fb[k]
             ctf_fb_k_t = ctf_fb_k.T
             sig_covar_coeff = ctf_fb_k @ covar_coeff @ ctf_fb_k_t
+
             sig_noise_covar_coeff = sig_covar_coeff + noise_covar_coeff
 
             mean_coeff_k = ctf_fb_k.apply(mean_coeff)
 
-            coeff_est_k = coeff_k - mean_coeff_k[:, np.newaxis]
-            coeff_est_k = sig_noise_covar_coeff.solve(coeff_est_k)
+            coeff_est_k = coeff_k - mean_coeff_k[np.newaxis, :]
+            coeff_est_k = sig_noise_covar_coeff.solve(coeff_est_k.T)
```

Comment:
No,  coeff_est_k is 2D. I think the traspose is required at this time, based on what `solve` is expecting.  I printed them below (before transpose) as an example.  Maybe `solve` needs to get changed...

```
------------------------------------------------------------ live log call -------------------------------------------------------------
2020/09/16 16:41:30 Expanding 2D image in a frequency-domain Fourier–Bessel basis using the fast method.
2020/09/16 16:41:30 eval_filters passed <class 'numpy.ndarray'> instead of Image instance
2020/09/16 16:41:30 eval_filters passed <class 'numpy.ndarray'> instead of Image instance
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
(5, 34) (5, 34)
(3, 34) (3, 34)
(2, 34) (2, 34)
(2, 34) (2, 34)
(5, 34) (5, 34)
(6, 34) (6, 34)
(9, 34) (9, 34)
PASSED
tests/test_covar2d.py::Cov2DTestCase::test07GetCWFCoeffsIdentityCTF 2020/09/16 16:41:30 Expanding 2D image in a frequency-domain Fourier–Bessel basis using the fast method.

------------------------------------------------------------ live log call -------------------------------------------------------------
2020/09/16 16:41:30 Expanding 2D image in a frequency-domain Fourier–Bessel basis using the fast method.
2020/09/16 16:41:30 eval_filters passed <class 'numpy.ndarray'> instead of Image instance
2020/09/16 16:41:30 eval_filters passed <class 'numpy.ndarray'> instead of Image instance
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
(32, 34) (32, 34)
PASSED
tests/test_covar2d.py::Cov2DTestCase::test08GetCWFCoeffsClean 2020/09/16 16:41:30 Expanding 2D image in a frequency-domain Fourier–Bessel basis using the fast method.

------------------------------------------------------------ live log call -------------------------------------------------------------
2020/09/16 16:41:30 Expanding 2D image in a frequency-domain Fourier–Bessel basis using the fast method.
2020/09/16 16:41:30 eval_filters passed <class 'numpy.ndarray'> instead of Image instance
2020/09/16 16:41:30 eval_filters passed <class 'numpy.ndarray'> instead of Image instance
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
2020/09/16 16:41:30 FFBBasis2D::evaluate_t passed numpy array instead of Image.
(32, 34) (32, 34)
PASSED
```

Reply:
Ok I see. Indeed, `solve` is “wrong”, but it gets tricky here. Naively, since the convention is row-major, we'd say that `solve` should broadcast along rows, which comes down to solving X A = B (X is K × M, A is M × N, and B is K × N). However, that's not the standard form of the equation, which is A X = B and implies a column-major convention. This is probably why `np.linalg.solve` also does the “wrong” thing. So let's just leave it.